central/auth/datastore/datastore_impl.go

```diff
@@ -0,0 +1,81 @@
+package datastore
+
+import (
+	"context"
+	"time"
+
+	"github.com/stackrox/rox/central/auth/datastore/internal/store"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+var (
+	_ DataStore = (*datastoreImpl)(nil)
+)
+
+type datastoreImpl struct {
+	store store.Store
+}
+
+func (d *datastoreImpl) GetAuthM2MConfig(ctx context.Context, id string) (*storage.AuthMachineToMachineConfig, bool, error) {
+	return d.store.Get(ctx, id)
+}
+
+func (d *datastoreImpl) ListAuthM2MConfigs(ctx context.Context) ([]*storage.AuthMachineToMachineConfig, error) {
+	return d.store.GetAll(ctx)
+}
+
+func (d *datastoreImpl) AddAuthM2MConfig(ctx context.Context, config *storage.AuthMachineToMachineConfig) (*storage.AuthMachineToMachineConfig, error) {
+	if config == nil {
+		return nil, errox.InvalidArgs.New("empty config given")
+	}
+	config.Id = uuid.NewV4().String()
+	if err := validateConfig(config); err != nil {
+		return nil, err
+	}
+
+	if err := d.store.Upsert(ctx, config); err != nil {
+		return nil, err
+	}
+	return config, nil
+}
+
+func (d *datastoreImpl) UpdateAuthM2MConfig(ctx context.Context, config *storage.AuthMachineToMachineConfig) (*storage.AuthMachineToMachineConfig, error) {
+	if err := validateConfig(config); err != nil {
+		return nil, err
+	}
+
+	if err := d.store.Upsert(ctx, config); err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+func (d *datastoreImpl) RemoveAuthM2MConfig(ctx context.Context, id string) error {
+	return d.store.Delete(ctx, id)
+}
+
+func validateConfig(config *storage.AuthMachineToMachineConfig) error {
+	if config.GetId() == "" {
+		return errox.InvalidArgs.New("empty ID given")
+	}
+
+	duration, err := time.ParseDuration(config.GetTokenExpirationDuration())
+	if err != nil {
+		return errox.InvalidArgs.New("invalid token expiration duration given").CausedBy(err)
+	}
+
+	if duration < time.Minute || duration > 24*time.Hour {
+		return errox.InvalidArgs.Newf("token expiration must be between 1 minute and 24 hours, but was %s",
+			duration.String())
+	}
+
```

Comment:
There are a couple examples (e.g. groups, imageintegrationhealth, networkgraph datastores), I don't mind moving this to the service layer though.

>Is it for sure customers wouldn't want token expiration shorter than 1 min or longer than 24 hours?
Fair question, the envisioned use-case is for CI, scanning images in under 1 minute is technically possible, but unlikely. Hence, the minimum of 1 minute.

For the upper-limit, the feature aims to provide a "short-lived" token, and we should be opinionated on the upper limit for that. For example, [AWS has a maximum of 36 hours](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_request.html#api_getsessiontoken) that you may set for expiration. I was debating setting the upper limit to 12 hours due to [the IdP tokenTTL](https://github.com/stackrox/stackrox/blob/cf825b3c19d0197ef4fe4e878bbe729f05fd6abc/pkg/auth/authproviders/provider_default_options.go#L44) (also see [ROX-9292](https://issues.redhat.com/browse/ROX-9292) for a recommended TTL of 12 hours), and ended up in the middle of our default TTL and the upper limit of AWS STS.

Reply:
I'm a strong believer in validation and business logic on the datastore level:
* service is not the only client of a datastore, there is also graphql
* if datastore is responsible for business logic, it can emit correct error types that increases consistency among its clients
* if store is responsible for low-level DB operations (i.e. now business logic), and service takes on business logic, why even having datastore abstraction?

Reply:
> there is also graphql

Is GraphQL planned to be a user of this particular datastore? When?

> why even having datastore abstraction?

It should not exist. It was useful at the time when Central was switchable between Rocksdb and Postgres.

The same reasoning can be turned back. If datastore is responsible for business logic, why have a service that simply proxies the calls?

My proposal was not "because I have an opinion and I want so", the other part of it was where to put tests. See https://github.com/stackrox/stackrox/pull/8145#discussion_r1363791796

Reply:
To be honest I'm a bit lost in the conversation: the tests have moved now to the service layer, and the datastore tests only have specifics (i.e. the foreign key behavior).

Is there anything required for _this_ PR to resolve this? The overall discussion seems more like a generic one to our whole concept of `service -> datastore -> store` which _may_ need rethinking.

Reply:
Yes, I'm happy how you moved tests. My last response was to @rukletsov .

----

Regarding

> to our whole concept of `service -> datastore -> store` which may need rethinking.

Please no.
Establishing guidelines for the entire team successfully failed multiple times or was tough and difficult.
If I can convince you, you can convince someone else and so on.

Reply:
Closing this since code changes are done and the discussion is stale.